### DISCUSSION TITLE
Desperate Plea for assistance: I got it working, but then tested H2OGPT. Bad idea! When back to PrivateGPT and can no longer build due to llama-cpp-python wheels

### DISCUSSION THREAD

**Discussion by csocha2** [imartinez/privateGPT#974] [Q&A]
Hi,
  This is a hail mary after spending countless hours on this. At this point, I'm desperate and sad because I can't get back to it working again. 

_Back story:_ After much effort, I finally got PrivateGPT working on my Windows 10 box. After enjoying it for a few days, I wanted to see how H2oGPT worked so I used their window installer. **BIG MISTAKE**. Not only could I not get H2OGPT to work on my local machine, I now cannot get PrivateGPT to run anymore, or install, or anything due to the llama-cpp-python wheels error. 

**I get the error as shown below. The steps I've taken to try to resolve it are as follows:**
- Uninstalled VS.NET 2022 Community
- Uninstalled MinGW GCC
- Updated PIP (already had the latest version)
- Rebooted
- Reinstalled VS.NET 2022 Community
    - Installed C++ CMake tools for Windows
    - Installed MS Build Tools 2022
    - There is nothing call Universal Windows Platform Development in Windows Installer. Assuming it comes built into Visual Studio.net 2022
- Reinstalled MinGW and selected:
    - ming32-gcc-g++ (The GNU C++ Compiler)  version 6.3.0-1
    - Also tried with mingw32-base
- Made sure c:\minGW\bin is in the system environment variable
- I can type gcc from the terminal in VS.Code. It can find it.
- I even tried to force it using, but same error: _pip install llama-cpp-python --force-reinstall --upgrade --no-cache-dir_
- _pip_ version is v3.10 and is up to date.
- I'm on an Intel 
- I have read tons of threads here and on google for this error no avail. It was working but now it's dead Jim.

**### The error I am getting is shown below:** 

Using cached Pygments-2.16.1-py3-none-any.whl (1.2 MB)
Using cached msoffcrypto_tool-5.1.1-py3-none-any.whl (34 kB)
**Building wheels for collected packages: llama-cpp-python
  Building wheel for llama-cpp-python (pyproject.toml) ... error
  error: subprocess-exited-with-error**

  × Building wheel for llama-cpp-python (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [131 lines of output]
     
     
      --------------------------------------------------------------------------------
      -- Trying 'Ninja (Visual Studio 17 2022 x64 v143)' generator
      --------------------------------
      ---------------------------
      ----------------------
      -----------------
      ------------
      -------
      --
      Not searching for unused variables given on the command line.
      CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
        Compatibility with CMake < 3.5 will be removed from a future version of
        CMake.
     
        Update the VERSION argument <min> value or use a ...<max> suffix to tell
        CMake that the project does not need compatibility with older versions.
     
     
      -- The C compiler identification is MSVC 19.37.32822.0
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx86/x64/cl.exe - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- The CXX compiler identification is MSVC 19.37.32822.0
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx86/x64/cl.exe - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Configuring done (4.0s)
      -- Generating done (0.0s)
      -- Build files have been written to: C:/Users/curti/AppData/Local/Temp/pip-install-vn0h_2dg/llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73/_cmake_test_compile/build
      --
      -------
      ------------
      -----------------
      ----------------------
      ---------------------------
      --------------------------------
      -- Trying 'Ninja (Visual Studio 17 2022 x64 v143)' generator - success
      --------------------------------------------------------------------------------
     
      Configuring Project
        Working directory:
          C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73\_skbuild\win-amd64-3.10\cmake-build
        Command:
          'C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\cmake\data\bin/cmake.exe' 'C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73' -G Ninja '-DCMAKE_MAKE_PROGRAM:FILEPATH=C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\ninja\data\bin\ninja' -D_SKBUILD_FORCE_MSVC=1930 --no-warn-unused-cli '-DCMAKE_INSTALL_PREFIX:PATH=C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73\_skbuild\win-amd64-3.10\cmake-install' -DPYTHON_VERSION_STRING:STRING=3.10.0 -DSKBUILD:INTERNAL=TRUE '-DCMAKE_MODULE_PATH:PATH=C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\skbuild\resources\cmake' '-DPYTHON_EXECUTABLE:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\python.exe' '-DPYTHON_INCLUDE_DIR:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\Include' '-DPYTHON_LIBRARY:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\libs\python310.lib' '-DPython_EXECUTABLE:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\python.exe' '-DPython_ROOT_DIR:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310' -DPython_FIND_REGISTRY:STRING=NEVER '-DPython_INCLUDE_DIR:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\Include' '-DPython_LIBRARY:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\libs\python310.lib' '-DPython3_EXECUTABLE:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\python.exe' '-DPython3_ROOT_DIR:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310' -DPython3_FIND_REGISTRY:STRING=NEVER '-DPython3_INCLUDE_DIR:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\Include' '-DPython3_LIBRARY:PATH=C:\Users\curti\AppData\Local\Programs\Python\Python310\libs\python310.lib' '-DCMAKE_MAKE_PROGRAM:FILEPATH=C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\ninja\data\bin\ninja' -DCMAKE_BUILD_TYPE:STRING=Release
     
      Not searching for unused variables given on the command line.
      -- The C compiler identification is MSVC 19.37.32822.0
      -- The CXX compiler identification is MSVC 19.37.32822.0
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx86/x64/cl.exe - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostx86/x64/cl.exe - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- Found Git: C:/Program Files/Git/cmd/git.exe (found version "2.34.1.windows.1")
      fatal: not a git repository (or any of the parent directories): .git
      fatal: not a git repository (or any of the parent directories): .git
      CMake Warning at vendor/llama.cpp/CMakeLists.txt:114 (message):
        Git repository not found; to enable automatic generation of build info,
        make sure Git is installed and the project is a Git repository.
     
     
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
      -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
      -- Looking for pthread_create in pthreads
      -- Looking for pthread_create in pthreads - not found
      -- Looking for pthread_create in pthread
      -- Looking for pthread_create in pthread - not found
      -- Found Threads: TRUE
      -- CMAKE_SYSTEM_PROCESSOR: AMD64
      -- x86 detected
      -- Configuring done (5.3s)
      -- Generating done (0.1s)
      -- Build files have been written to: C:/Users/curti/AppData/Local/Temp/pip-install-vn0h_2dg/llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73/_skbuild/win-amd64-3.10/cmake-build
      [1/7] Building C object vendor\llama.cpp\CMakeFiles\ggml.dir\k_quants.c.obj
      [2/7] Building C object vendor\llama.cpp\CMakeFiles\ggml.dir\ggml.c.obj
      FAILED: vendor/llama.cpp/CMakeFiles/ggml.dir/ggml.c.obj
      C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1437~1.328\bin\Hostx86\x64\cl.exe  /nologo -DGGML_USE_K_QUANTS -D_CRT_SECURE_NO_WARNINGS -IC:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73\vendor\llama.cpp\. /DWIN32 /D_WINDOWS /O2 /Ob2 /DNDEBUG -std:c11 -MD /arch:AVX2 /showIncludes /Fovendor\llama.cpp\CMakeFiles\ggml.dir\ggml.c.obj /Fdvendor\llama.cpp\CMakeFiles\ggml.dir\ /FS -c C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73\vendor\llama.cpp\ggml.c
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\winbase.h(8816): warning C5105: macro expansion producing 'defined' has undefined behavior
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(473): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\shared\wtypes.h(742): note: in expansion of macro '_VARIANT_BOOL' 
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(473): error C2059: syntax error: '/'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(488): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\shared\wtypes.h(742): note: in expansion of macro '_VARIANT_BOOL' 
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(488): error C2059: syntax error: '/'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(516): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(517): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(519): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(520): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(651): error C2079: 'varDefaultValue' uses undefined struct 'tagVARIANT'
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(936): error C2079: 'varValue' uses undefined struct 'tagVARIANT'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(318): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\shared\wtypes.h(742): note: in expansion of macro '_VARIANT_BOOL' 
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(318): error C2059: syntax error: '/'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(377): error C2371: 'pvarVal': redefinition; different basic types
      C:\Program Files (x86)\Windows Kits\8.1\include\um\oaidl.h(496): note: see declaration of 'pvarVal'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(378): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(379): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(382): error C2059: syntax error: '}'
      C:\Program Files (x86)\Windows Kits\8.1\\include\\\um\propidl.h(383): error C2059: syntax error: '}'
      [3/7] Building CXX object vendor\llama.cpp\CMakeFiles\llama.dir\llama.cpp.obj
      ninja: build stopped: subcommand failed.
      Traceback (most recent call last):
        File "C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\skbuild\setuptools_wrap.py", line 674, in setup
          cmkr.make(make_args, install_target=cmake_install_target, env=env)
        File "C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\skbuild\cmaker.py", line 697, in make
          self.make_impl(clargs=clargs, config=config, source_dir=source_dir, install_target=install_target, env=env)      
        File "C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\skbuild\cmaker.py", line 742, in make_impl
          raise SKBuildError(msg)
     
      An error occurred while building with CMake.
        Command:
          'C:\Users\curti\AppData\Local\Temp\pip-build-env-j4g86xr1\overlay\Lib\site-packages\cmake\data\bin/cmake.exe' --build . --target install --config Release --
        Install target:
          install
        Source directory:
          C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73
        Working directory:
          C:\Users\curti\AppData\Local\Temp\pip-install-vn0h_2dg\llama-cpp-python_facf68fc7af54f4ab3c9531685dbfd73\_skbuild\win-amd64-3.10\cmake-build
      Please check the install target is valid and see CMake's output for more information.
     
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for llama-cpp-python
Failed to build llama-cpp-python

**Comment by csocha2** [comment]
Update: 
I found a thread about popping up Visual Studio Installer and launching the Visual Studio Build Tools 2022. I then navigated to the folder I had PrivateGPT in and it successfully built the llama.cpp.python file. 

I noticed that when I do a c:\PATH using the Select Developer Command Prompt for VS 2022 , I get a totally different set of paths than if I do it in VS.CODE terminal, so clearly something is up with the path discrepancy all things being equal. Not specifically what yet.

Can verify: This fixed it.

**Reply by TomLucidor** [reply]
Wait, can you detail more on this?
